ddprof-lib/src/main/cpp/livenessTracker.cpp

```diff
@@ -263,13 +270,20 @@ void LivenessTracker::track(JNIEnv* env, AllocEvent &event, jint tid, jobject ob
         return;
     }
 
+    static thread_local std::mt19937 gen(std::random_device{}());
+    static thread_local std::uniform_real_distribution<> dis(0, 1.0);
+    static thread_local double skipped = 0;
+
+    if (_subsample_ratio < 1.0 && dis(gen) > _subsample_ratio) {
+        skipped += static_cast<double>(event._weight) * event._size;
+        return;
+    }
+
     jweak ref = env->NewWeakGlobalRef(object);
```

Comment:
`object` would never be null.

The check for null on the weakref is just a safety net - if we get null there it means we are completely out of memory and the system will go down very soon (hopefully nobody is trying to recover from `OutOfMemoryError`)

Technically, the interval the object represents should be included in the skipped bytes but I have doubts that it would make a noticeable difference considering the most likely terminal phase of JVM we are observing there.